src/cargo/core/resolver/resolve.rs

```diff
@@ -19,21 +18,47 @@ use super::encode::Metadata;
 /// for each package.
 #[derive(PartialEq)]
 pub struct Resolve {
-    pub graph: Graph<PackageId>,
-    pub replacements: HashMap<PackageId, PackageId>,
-    pub empty_features: HashSet<String>,
-    pub features: HashMap<PackageId, HashSet<String>>,
-    pub checksums: HashMap<PackageId, Option<String>>,
-    pub metadata: Metadata,
-    pub unused_patches: Vec<PackageId>,
+    graph: Graph<PackageId>,
+    dependencies: HashMap<(PackageId, PackageId), Vec<Dependency>>,
```

Comment:
Hm, naively I would think that edge data would be stored as, well, edges in the graph. 

Perhaps we should refactor `pub struct Graph<N>` to

```
pub struct Graph<V, E> {
   edges: HashMap<V, HashMap<V, E>>
}
```

and instantiate it as `Graph<PackageId, Vec<Dependency>>`.

If we go down that road, then perhaps let's explore using `petgraph` for it? rust-lang/rust already depends on it, so it won't be a too new dependency.

Reply:
>The problem is that lock files are then a different graph data structure (no Dependency information) than the Resolve itself (which has Dependency information). That turned out to be a much larger refactoring than necessary so I instead kept the edge data in a side table

Ok, though I don't feel that lockfiles should be a particular problem? For lockfiles, we could store the empty `Vec` of edges? I.e., [this line](https://github.com/rust-lang/cargo/blob/master/src/cargo/core/resolver/encode.rs#L109) should be changed from `g.link(id.clone(), to_depend_on);` to

```
// Note: we don't store edges info in the lockfile, so using an empty Vec here
g.link(id.clone(), to_depend_on, Vec::new());
```

That might cause problems elsewhere of course :)

Reply:
Hm, let's merge this implementation, and I'll look into probably petgraph separately?